core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -443,9 +444,19 @@ public Optional<TaskId> getAttachedTaskId()
             return Optional.ofNullable(this.taskId.get());
         }
 
+        @Override
+        public void setMemoryRequirement(DataSize memoryRequirement)
+        {
+            long newBytes = memoryRequirement.toBytes();
+            long previousBytes = memoryLease.getAndSet(newBytes);
+            if (newBytes < previousBytes) {
+                wakeupProcessPendingAcquires();
+            }
+        }
+
         public long getMemoryLease()
         {
-            return memoryLease;
+            return memoryLease.get();
```

Comment:
> does BinPackingSimulation support memory lease not being constant.

It may result that bin-packing decision is not 100% accurate but even without that it is still the case. We are already operating in dynamic environment when runtime memory usage of tasks (which is also taken into account) is changing.
Yet still I will smooth some cases you pointed out offline, so race conditions are less likely.